### ISSUE TITLE
Selfservice account deletion 

### ISSUE THREAD

**Issue by zepatrik** [ory/kratos#596]
**Is your feature request related to a problem? Please describe.**

Users have to be able to delete their account including all personal data (GDPR/... complaint).

**Describe the solution you'd like**

The new flow `/self-service/flows/delete` should allow users to delete all data associated with them. This hast to be at least CSRF protected, we have to check whether we want to (and are allowed to) reauthenticate a user (similar to password change).

**Describe alternatives you've considered**

Delete the internet.

**Comment by zepatrik** [comment]
Not really, this operation needs some checks, similar to the password update, where the session has to be recent enough.

**Comment by wenqi73** [comment]
@zepatrik Are there any plans for this feature?

**Comment by zepatrik** [comment]
Not really, all you can see here. But this is definitely an important issue.

**Comment by Quasarman** [comment]
How would I go about allowing users to delete their account without this feature?

**Comment by zepatrik** [comment]
You could use the admin API, and of course need some back-end wrapper around that as you should not expose it.

**Comment by Quasarman** [comment]
Alright thank you.

**Comment by alantbarlow** [comment]
Im just going over the documentation of this, and it seems like this could be accomplished by simply allowing the user to delete his own database record from within the authentication database only. Then once the user confirms the deletion, it could activate "hooks" (similar to the other ones) that run before and after being deleted from the system.

This should allow the user to self-delete their identity and allow the web hook delete all other data related to that userID. Correct me if I'm wrong but I don't think this would be too hard to implement.

**Comment by woylie** [comment]
While a self-service account deletion flow can be useful in some cases, in more complex setups it wouldn't suffice. Think of a distributed system where user data must be deleted or pseudonymised on multiple services / from multiple databases, not just Kratos. In those cases I would rather track and handle deletion requests outside of Kratos and handle the Kratos part via the admin API. Just throwing this in to say that in case this will be added in a future version, you should be able to disable this feature in the config.

**Comment by openscript** [comment]
I don't agree with @woylie. As the self creation of and identity, the self deletion of an entity should be handled by Kratos. There should be a hook, so other systems are notified about the deletion.

**Comment by woylie** [comment]
> I don't agree with @woylie. As the self creation of and identity, the self deletion of an entity should be handled by Kratos. There should be a hook, so other systems are notified about the deletion.

There are cases where self-service account deletion is undesired. Think of organization accounts, where user accounts are managed by an administrator.

**Comment by openscript** [comment]
Yes sure, like self registration it should be configurable as deactivated.

**Comment by kmherrmann** [comment]
I think a self-service account deletion hardly suffers for GDPR compliance - you have to make sure data is cleared from your other systems as well in almost all cases. Thus our recommendation to implement your own endpoint to orchestrate actions, and call the Admin Delete endpoint. This way you can also choose to instead remove credentials and PII and keep a skeleton record as needed.

The self-service endpoint with Actions/Webhooks is a valid approach, too. As Aeneas mentions, this isn't on our near-term roadmap, though. If you'd like to contribute it, we'd be happy to review! Important is that any such functionality includes an explicit confirmation step with CSRF protection.

**Comment by janoschp** [comment]
Any progress on this? Whether it is a soft or hard delete should be up to the developer and ory should offer this flow.

**Comment by aeneasr** [comment]
This is our current standpoint:

> I think a self-service account deletion hardly suffers for GDPR compliance - you have to make sure data is cleared from your other systems as well in almost all cases. Thus our recommendation to implement your own endpoint to orchestrate actions, and call the Admin Delete endpoint. This way you can also choose to instead remove credentials and PII and keep a skeleton record as needed.
> 
> The self-service endpoint with Actions/Webhooks is a valid approach, too. As Aeneas mentions, this isn't on our near-term roadmap, though. If you'd like to contribute it, we'd be happy to review! Important is that any such functionality includes an explicit confirmation step with CSRF protection.

GDPR deletion has impact on many areas - you need to clean data from CRM, back ups, emails, and other storage. This is more involved than just "delete my account". We're not sure how to properly do this at this time and recommend you to build your own account deletion using the admin identity delete endpoint.
